### DISCUSSION TITLE
PKIX path building failed - SunCertPathBuilderException when using IKVM.Maven.Sdk with Docker for Windows

### DISCUSSION THREAD

**Discussion by MattBayliss** [ikvm-revived/ikvm-sdk-maven#38] [Q&A] [answered]
I'm newish to Docker, IKVM and IKVM.Maven.SDK, and I'm trying to create a docker container for a dotnet project that uses IKVM.Maven.Sdk, but I get the following error trying to publish my dotnet project:
```
C:\Users\ContainerAdministrator\.nuget\packages\ikvm.maven.sdk\1.5.5\buildTransitive\netstandard2.0\IKVM.Maven.Sdk.targets(96,9): 
warning : Could not transfer artifact commons-io:commons-io:jar:2.13.0 from/to central (https://repo1.maven.org/maven2/): 
sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: 
unable to find valid certification path to requested target [C:\TikaTest\TikaTest.csproj]
```

How do I solve these certificate issues? My DuckDuckGo-ing shows that the solution for a Java project is to copy a local cacerts keyfile into the container, with something like ([from this blog](https://betterjavacode.com/programming/how-to-resolve-pkix-path-building-failed-in-docker-container)):

```
COPY ./cacerts /usr/lib/jvm/java-1.8.0-amazon-corretto.x86_64/jre/lib/security/cacerts
```

Is there an equivalent for my Windows Docker container with IKVM/IKVM-Maven?

Here's my very basic Dockerfile - stripped down to just a dotnet restore + publish to troubleshoot the issue:

```docker
FROM mcr.microsoft.com/dotnet/sdk:6.0-windowsservercore-ltsc2022 AS build
WORKDIR /TikaTest

COPY . ./

RUN dotnet restore
RUN dotnet publish -c Release -o out
```

And here's the dotnet project I'm using for my troubleshooting: https://github.com/MattBayliss/TikaTest-IKVM.Maven.Sdk

Thanks for any assistance!

Matt.

**Comment by wasabii** [comment]
Well..... I don't know much about the type of process you're using. But it looks like you're trying to run the actual build inside of a Container? So the container needs the full SDK installed, and it needs internet access, and the capability to run Maven, and internet access, and everything else?

That seems... possible. But a pain in the ass. Run the build outside of the container, and copy the results into it.

**Comment by wasabii** [comment]
That said, IKVM derives certificates from .NET.

So, whatever is available in StoreLocation.LocalMachine and StoreLocation.CurrentUser for X509Store. No idea how that plays out inside of a Windows container.

**Comment by MattBayliss** [comment]
Sorry for the delayed response - thank you for your replies! I've been using your suggestions to try and get things working. Basically I'm trying to get a text extracting component of a larger C# .Net product built and deployed using Azure Pipelines in devops to end up in our Azure Container Registry, to Azure Kubernetes Service. I'm using Maven to get Tika working (you helped me on the TikaOnDotNet project too). All of the other components/services are built as part of the docker image creation in Azure Pipelines.

It has internet access - it's restoring Nuget packages (like IKVM.Maven.SDK)

I've tried manually importing certificates by exporting them from my browser and doing commands like:
```
RUN powershell Import-Certificate -FilePath .\repo1-maven-org-chain.pem -CertStoreLocation Cert:\CurrentUser\My
```

For each repo error, but no change.

I tried installing JDK 8.1 and Maven to the container, but same issues.

Going to try with a simple Java Maven project, to see if I get the same cert issues, but [this StackOverflow](https://stackoverflow.com/questions/45612814/maven-error-pkix-path-building-failed-unable-to-find-valid-certification-path#60156232) suggests running:

```
mvn clean package -Dmaven.wagon.http.ssl.insecure=true -Dmaven.wagon.http.ssl.allowall=true
```

Which I can try with my Java attempt, if it fails, against my pom file. Will post any new details as they come to light. Thanks for your time.

**Comment by wasabii** [comment, marked as answer]
My best suggestion is to simply not do the build INSIDE of a container. Build it outside, and copy the results in.
